src/recordlinker/database/mpi_service.py

```diff
@@ -640,48 +626,43 @@ def generate_non_match_tuning_samples(
     if repeat_probability >= 0.5:
         raise ValueError("Too many pairs requested for sample size")
 
-    # Start with a large sub-sample from which we'll draw pairs
-    query = expression.select(
-        models.Patient.id,
-        models.Patient.person_id,
-        models.Patient.data
-    ).where(
-        ~models.Patient.person_id.is_(None)
-    ).order_by(
-        get_random_function(session.get_bind().dialect)
-    ).limit(
-        sample_size
+    random_ids: list[int] = [
+        row[0]
+        for row in session.query(models.Patient.id)
+        .order_by(get_random_function(session.get_bind().dialect))
+        .limit(sample_size)
+        .all()
+    ]
+    query: expression.Select = (
+        select(models.Patient.id, models.Patient.person_id, models.Patient.data)
+        .where(models.Patient.id.in_(bindparam("ids")))
```

Comment:
```suggestion
       .where(models.Patient.id.in_(bindparam("ids", expanding=True)))
```
I _think_ you need to add `expanding=True` or else it might treat the whole list of ids as a single entity rather than checking if the Patient.id is in the list. See documentation [here](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.bindparam.params.expanding)

Reply:
I believe its working, but yeah the documentation def says its required, so I'll merge that in.  Great find!